### ISSUE TITLE
Loyc += Flame?

### ISSUE THREAD

**Issue by qwertie** [jonathanvdc/Flame#29]
So you might have noticed I made an "organization" called loycnet (because some inactive, zero-contribution user is sitting on "loyc"). The [ecsharp](https://github.com/loycnet/ecsharp) repo there will become the main one for the source code (and maybe for ecsharp.net too).

And I'm rewriting the front page of loyc.net. I want to make Loyc more inclusive and start building some kind of community, not that I know how to recruit. But an obvious place to start is by listing Flame as a Loyc project on the home page. Does that sound good to you? How would you describe it in one paragraph? I guess Flame doesn't have its own home page, so how about I link to [here](https://github.com/jonathanvdc/Flame/)?

BTW I made a thread for [general conversation](https://github.com/loycnet/ecsharp/issues/1) which we could use to contact each other about minor questions and random thoughts. Care to subscribe/watch it?

**Comment by qwertie** [comment]
Oh, and of course, eventually I'd like to merge ecsc into the ecsharp repo. And I'd like to do a big code review of Flame and probably lob some complaints, er, suggestions your way. How does that sound?

**Comment by jonathanvdc** [comment]
I'd be absolutely delighted if you listed Flame on loyc.net. Thanks! 

I feel like describing Flame in one paragraph is actually a bit of a challenge (though I suppose that is true for most software systems). Let me give it a try:

Flame is a set of libraries focused on generating code for managed languages. For .NET developers, Flame can be likened to LINQ expression trees: they both abstract away the underlying details of the platform you're generating code for, and leave you with the relatively straightforward task of converting your source code to a tree-based intermediate representation. The main difference between Flame and expression trees is that Flame is intended to be used in an ahead-of-time compilation context, whereas expression trees generate code at run-time. Flame also has a few neat tricks up its sleeve to make building a compiler easier: 
* it can target multiple platforms, 
* it can optimize the IR your front-end produces,
* it can save, load and statically link Flame IR files,
* and it can handle the command-line interface.

I'm not sure if that counts as one paragraph. Maybe you'll want to shorten or re-word (parts of) it.

**Comment by jonathanvdc** [comment]
A code review of Flame sounds great. Your ~~complaints~~ suggestions would be most welcome. Some parts of Flame are an absolute mess, mostly because I didn't have enough time to get things right the first time around (Flame is a fairly big project with lots of features, and there's always something to do; I have at times used hacks to make progress). Thanks for the offer!

**Comment by jonathanvdc** [comment]
Well, I'm not sure if `ecsc` should be merged into `ecsharp`, mostly for usability reasons. People are always complaining about how hard it is to get the LLVM monorepo working, and I think that it'd be a shame if the `ecsharp` repo suffered a similar fate. There's no circular dependency between `ecsharp` and `ecsc`, so why merge them?

On the other hand, I certainly wouldn't mind migrating `ecsc` to loycnet, though. Would you like me to do that?

**Comment by jonathanvdc** [comment]
Oh, BTW. If at some point you feel the need to `git clone` Flame, then I recommend that you don't clone the entire `git` history. I screwed up at some point by accidentally committing and deleting some large files, which has since bloated the `git` history. A full clone takes really long now.

TL;DR: use `--depth=n` to specify a shallow clone. For example:

```sh
$ git clone --depth=50 https://github.com/jonathanvdc/Flame
```

**Comment by qwertie** [comment]
Once ecsc gets to a usable state (notably: it should be self-hosting) the status quo of ecsharp not being a true EC# compiler becomes confusing. And ecsc is not a large project (it's separate from Flame, right?) so it could be merged in. What makes you think it would be hard to get the ecsharp repo 'working'?

I guess there would be a seeming circular repo-dependency where Flame uses LoycCore which is embedded in ecsharp, but ecsc (now in ecsharp) depends on Flame. Especially now that I can't push to LoycCore anymore (as `git subtree push` stopped working for reasons unknown, a few months ago). But of course this is not a real circular dependency; it's arguably not even as bad as the existing circular dependency where LLLPG depends on EC# and LoycCore while EC# and LoycCore both depend on LLLPG binaries. (And of course eventually ecsc will depend on itself - the inevitable fate of compilers written in their own language...)

**Comment by jonathanvdc** [comment]
My original thinking was that an `ecsharp`/`ecsc` would make automated testing for Flame a slower and more involved process when I move to EC# as the implementation language for the EC# core libraries. Right now, I just build Flame, which includes `dsc`, and run the automated tests. 

Switching to a separate `ecsc` as the compiler for both Flame and the test framework would require me to first build Flame, then build the `ecsc` repo, and finally run the tests. Bundling `ecsc` with `ecsharp` would add building the other tools in the `ecsharp` repository to that list. 

However, having thought about it for a couple of days, I have come to the conclusion that I was wrong to oppose merging `ecsc` into the `ecsharp` repository. The main pain in switching to `ecsc` is cloning and building `ecsc`; whether that is part of the `ecsharp` repository may make that slightly more difficult, but that would actually be the least of my headaches.

Also, I can understand why you want to merge `ecsc` into the `ecsharp` repository once `ecsc` matures: it will certainly make it easier to keep `ecsc` in sync with the EC# parser. So consider me on board with the plan to merge `ecsc` into `ecsharp` someday.

**Comment by qwertie** [comment]
Your directory structure doesn't categorize the folders, i.e. which back-ends and front-ends are not immediately distinguishable. On Loyc.net I'd like to give two examples of back-ends. Which ones are most stable?

**Comment by qwertie** [comment]
> it can target multiple platforms

With the same source code? How much of the BCL can one use when not targeting .NET? e.g. are `System.String` and `List<T>` fully implemented?

**Comment by qwertie** [comment]
> Bundling ecsc with ecsharp would add building the other tools in the ecsharp repository to that list. 

Shouldn't be too bad. A full rebuild of ecsharp does take 17 seconds on my laptop, but that includes Baadia and LLLPG which you can probably exclude (remind me to get rid of Baadia one of these days... really I want to rewrite it in Javascript anyhow so it will run on the web), and hopefully most of your builds are incremental.

**Comment by jonathanvdc** [comment]
The only stable back-ends that implement all features are the CLR back-end and the on-disk IR back-end. Flame's _architecture_ is not tied to a single platform, but I don't really have the time to maintain the other back-ends properly. 

If you want to mention another back-end, then I would recommend listing the WebAssembly back-end. It does not implement a number of essential features &ndash; reference types, generics and vtables, mostly &ndash; and targets the outdated WebAssembly S-expression format, but it doesn't have the technical debt that plagues the C++ and Python back-ends.

**Comment by jonathanvdc** [comment]
> > it can target multiple platforms

> With the same source code? How much of the BCL can one use when not targeting .NET? e.g. are `System.String` and `List<T>` fully implemented?

Sadly, no. Flame doesn't, and has no ambitions to, translate BCL classes like `List<T>` to target platform classes. The BCL can only be used when you set the .NET platform as the runtime by using the following compiler switch: `-runtime clr`. (By the way, the runtime is distinct from the target platform; it is perfectly possible to reference BCL classes when compiling for, say, Flame IR, by using `-platform ir -runtime clr`.)

However, I do plan on writing a cross-platform standard library eventually, which Flame-based compilers can then reference by default when compiling for certain platforms, and it's not unthinkable for there to be a `List<T>` in that library.

I guess that's sort of similar to the multi-language standard library (MLSL) idea you sometimes mention in your blog posts. I'd love to work with you on that, if you're okay with it.

**Comment by qwertie** [comment]
How about this blurb: [Flame](http://github.com/jonathanvdc/flame) is a set of "middle-end" components designed to help write "managed" (C#-like) compilers. It has is own IR (separate from Loyc trees) which it can optimize, and it supports static linking of Flame IR files. The CLR backend is the only stable one; a WebAssembly backend is in early stages.

**Comment by jonathanvdc** [comment]
That's an awesome description! Though, if I may make a suggestion: why don't we just lose the quotes? Those might give the impression that we're making things up on the spot &ndash; we're not; we didn't invent middle-ends or managed languages! How about italicizing _middle-end_ and _managed,_ or replacing them with hyperlinks? In my brief search, I haven't found a definition for _managed code_ that doesn't refer specifically to code that runs on the CLR, but I think https://en.wikipedia.org/wiki/Compiler#Three_phases_compiler_structure might be an appropriate place to refer users to for more information on what a middle-end is, exactly.

**Comment by jonathanvdc** [comment]
By the way, I started working on [a tutorial on building a compiler with Flame](https://github.com/jonathanvdc/objective-kaleidoscope) (inspired by LLVM's [kaleidoscope tutorial](http://llvm.org/docs/tutorial/)) this summer, but I sort of ran out of time halfway through building the AST and the parser. 

I'm considering finishing that now. Do you think that people might be interested in reading a tutorial on how to use Flame? Actually, let me phrase that differently: would you be interested in reading a tutorial on building a Flame-based compiler? If so, which artifacts are you most interested in? The code or the tutorial itself? My plan was to talk about the interesting bits of the code in the tutorial, and then allow the reader to explore the details of the compiler by browsing through the source code at their leisure. 

I mostly just want to

1. inform people that Flame is a thing, and
2. show them how a basic Flame-based compiler works.

Does that sound like a decent approach?

**Comment by qwertie** [comment]
> why don't we just lose the quotes?

Well, middle-end sounds funny to me because by definition, it's not an "end", it's the "middle". So I used quotes because I did not perceive the term to be in common use. And I put "managed" in quotes because the term is usually used as a synonym for the CLR, which in this case it is not. But if you still think I should use italics, I will.

Yes, I'd be interested in reading your tutorial. I do not enjoy just jumping in and figuring stuff out - that may be why I write so much documentation, I'm just trying to follow the golden rule.

> inform people that Flame is a thing, and

Yes, uh, I'd like to hear the history of it (why did you make it?), what are your goals, and what makes Flame different from other projects.

> show them how a basic Flame-based compiler works.

I am interested most in learning the architecture of the system first, but using example code and concrete class names will help you to communicate that effectively. The second thing I want to know is how one would write a compiler, and for that you should provide a complete working example: a compiler that is too small to use for real programming, but nontrivial enough to show the most important features. Certainly one should be able to do `(.new List!Int32()).Add(3)` in this language, for instance.

I'm thinking you should use LESv3 as the syntax, although it's not fully stable - see [chat](https://github.com/loycnet/ecsharp/issues/1).
